Stats auto refresh

max-hoffman opened the pull request:
Three new system variables to configure stats auto refresh, one for turning on, one for a timer of how often to check whether stats are fresh, and a third for a freshness threshold to trigger writing new stats. The system variables are global, and will apply to all databases in a server context. Adding or dropping databases, tables, and indexes are reflected by the statistics provider.

Statistic updates now funnel through a process where we preserve buckets for pre-existing chunks and create new buckets specifically for new chunk ordinals. The auto-refresh threshold compares the number of new and deleted buckets compared the the existing bucket count. So inserts, updates, and deletes all create new chunks or delete chunks relative to the previously saved statistics, and the fraction of (new+deleted)/previous count is compared against the threshold.

TODO:
- how much logging is appropriate
- go server tests
- is stats dataset replication blocked by default?

other:
- ANALYZE table and auto-refresh concurrency, last write wins? as long as auto-refresh thread doesn't die
- drop database/ANALYZE concurrency, error for writing stats to a deleted folder

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>340e43c</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>340e43c</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>340e43c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f180d20</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>f180d20</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f180d20</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>97d8a04</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>97d8a04</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>97d8a04</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>81dba2b</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>81dba2b</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>81dba2b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

coffeegoddd commented:
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>5edf046</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>5edf046</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>5edf046</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998787 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>38392dd</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>38392dd</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>38392dd</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

nicktobey reviewed (commented):
Should we have bats tests for setting different values of threshold and interval?

nicktobey commented on go/libraries/doltcore/sqle/stats/auto_refresh.go:
```diff
@@ -0,0 +1,303 @@
+						}
+						curCnt := float64(len(curStat.active))
+						updateCnt := float64(len(idxMeta.updateChunks))
+						deleteCnt := float64(len(curStat.active) - len(idxMeta.preexisting))
```
Should this be reversed? The number deleted is the pre-existing minus the current count?

max-hoffman replied:
the naming here is somewhat rough and ready, but `curStat` is the existing statistics on disk, `idxMeta` is the update metadata. `idxMeta.preexisting` is the fraction of chunks in the update metadata overlapping with `curStat`'s chunks. Will change `idxMeta` to indicate it's a candidate update

nicktobey commented on go/libraries/doltcore/sqle/stats/auto_refresh.go:
```diff
@@ -0,0 +1,303 @@
+	var missingOffsets [][]uint64
+	var offset uint64
+	for _, n := range levelNodes {
+		// check if hash is in current list
```
Can you add a comment explaining what's going on in this block?

max-hoffman replied:
Added a comment. Basically partition the target level histogram hashes for the current tree into 1) preserved or 2) missing, and track scanning metadata for missing chunks

nicktobey replied:
Okay I follow. Is there a particular reason we do it this way by storing all the hashes at some level of the tree, instead of just storing the root hash?

max-hoffman replied:
Maybe I'm misunderstanding the question, but each histogram bucket is the contexts of a chunk at the target level, one-to-one equivalence. We do track the table hash to short-circuit the target level comparison if nothing has changed, but otherwise we use structural sharing to 1) get an estimate of how much has changed, and 2) avoid unnecessarily recomputing buckets that haven't changed.

nicktobey replied:
Ah, okay. I missed that the chunks being tracked here mapped 1-1 with histogram buckets.

This makes sense.

max-hoffman replied:
WIil update docstrings to clarify

nicktobey commented on go/libraries/doltcore/sqle/stats/stats_provider.go:
```diff
@@ -26,17 +27,20 @@ import (
 type DoltStats struct {
 	level         int
 	chunks        []hash.Hash
+	active        map[hash.Hash]int
```
Can you add docstrings explaining the purpose of these fields? `active` in particular is unclear to me.

nicktobey commented on go/libraries/doltcore/sqle/stats/stats_provider.go:
```diff
@@ -142,48 +172,55 @@ func (s DoltHistogram) toSql() []*stats.Bucket {
 // Init scans the statistics tables, populating the |stats| attribute.
 // Statistics are not available for reading until we've finished loading.
 func (p *Provider) Load(ctx *sql.Context, dbs []dsess.SqlDatabase) error {
-	p.mu.Lock()
```
Is it okay that this doesn't have a lock anymore?

max-hoffman replied:
rewrote how the concurrency control works, it's fairly straightforward but i might need to add some go tests for race detection

nicktobey commented on go/libraries/doltcore/sqle/stats/write.go:
```diff
@@ -139,3 +140,54 @@ func flushStats(ctx *sql.Context, prev prolly.Map, tableStats map[sql.StatQualif
+		keyBuilder.PutString(0, qual.Database)
+		keyBuilder.PutString(1, qual.Table())
+		keyBuilder.PutString(2, qual.Index())
+		keyBuilder.PutInt64(3, 10000)
```
Add a comment for which column is being written here, and why it gets a hardcoded 10k.

max-hoffman replied:
added a `maxBucketFanout` constant to make clearer

nicktobey commented on go/libraries/doltcore/sqle/stats/update.go:
```diff
@@ -200,6 +174,58 @@ func refreshStats(ctx *sql.Context, indexes []sql.Index, idxMetas []indexMeta) (
+		if i < len(oldHist) && oldHist[i].Chunk == chunkAddr {
+			mergeHist = append(mergeHist, oldHist[i])
+			i++
+		} else if j < len(newHist) && newHist[j].Chunk == chunkAddr {
```
Possible bug: if an identical chunk appears in both the old and new histograms, then j will never be incremented and all the new chunks that come after it will be skipped.

max-hoffman replied:
yes this was a bug

nicktobey commented on go/libraries/doltcore/sqle/stats/update.go:
```diff
@@ -200,6 +174,58 @@ func refreshStats(ctx *sql.Context, indexes []sql.Index, idxMetas []indexMeta) (
+func firstRowForIndex(ctx *sql.Context, prollyMap prolly.Map, keyBuilder *val.TupleBuilder, prefixLen int) (sql.Row, error) {
+	buffPool := prollyMap.NodeStore().Pool()
+
+	firstIter, err := prollyMap.IterOrdinalRange(ctx, 0, 1)
```
Why are these values of start and stop used?

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999074 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>b5c93ad</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>b5c93ad</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>b5c93ad</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

max-hoffman commented on go/libraries/doltcore/sqle/enginetest/stats_queries.go:
```diff
@@ -368,18 +368,37 @@ var DoltStatsIOTests = []queries.ScriptTest{
 		},
 	},
+	{
+		Name: "stats updates",
```
tested incrementally rewriting stats depending on which chunks were edited, rather than a full scan. changed test name

max-hoffman commented:
I'm going to try to get some sort of tests that are race sensitive,  which might be a better format for adding finer-grained parameter testing.

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>7198173</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>7198173</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>7198173</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>bbfbc36</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>bbfbc36</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>bbfbc36</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

coffeegoddd commented:
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>18fd075</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>18fd075</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>18fd075</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

max-hoffman commented:
@nicktobey I added some functions for controlling auto refresh and some minimalist testing to make sure they stop, restart, and drop stats. The intention is to give a bit more control when things go wrong. I had to refactor some of the stats provider lifecycle so that background threads and sessions have consistent root value views. I also added concurrency testing for some common cases. There's probably a lot more testing we could add, but I think this is a comfortable MVP, things mostly work and when they don't we mostly won't brick the database. Might be helpful to check `TestStatsAutoRefreshConcurrency`, `enginetest.StatProcTests`, and `dprocedures/stat_funcs.go`.

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>be41ce9</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>be41ce9</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>be41ce9</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

nicktobey commented on go/libraries/doltcore/sqle/stats/stats_provider.go:
```diff
@@ -317,10 +456,14 @@ func (p *Provider) RowCount(ctx *sql.Context, db, table string) (uint64, error)
+			if strings.EqualFold(db, qual.Database) && strings.EqualFold(table, qual.Table()) {
+				if s.AvgSize > avgSize {
+					avgSize = s.AvgSize
+				}
```
This should probably be returning avgSize, not 0.

nicktobey commented on go/libraries/doltcore/sqle/stats/stats_provider.go:
```diff
@@ -317,10 +456,14 @@ func (p *Provider) RowCount(ctx *sql.Context, db, table string) (uint64, error)
+		for qual, s := range dbStat.stats {
+			if strings.EqualFold(db, qual.Database) && strings.EqualFold(table, qual.Table()) {
+				if s.AvgSize > avgSize {
+					avgSize = s.AvgSize
```
Why do we take the highest value for avgSize across all stats?

max-hoffman replied:
Changed this to match the primary index

nicktobey commented on go/libraries/doltcore/sqle/stats/write.go:
```diff
@@ -24,18 +24,26 @@ import (
 	"github.com/dolthub/dolt/go/store/val"
 )
 
+const maxBucketFanout = 200 * 200
```
Add a comment explaining why this value was chosen.

max-hoffman commented on go/libraries/doltcore/sqle/dprocedures/stats_funcs.go:
```diff
@@ -0,0 +1,123 @@
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+)
+
+func statsFunc(name string) func(ctx *sql.Context, args ...string) (sql.RowIter, error) {
```
yep good idea

coffeegoddd commented:
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f09626b</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>f09626b</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f09626b</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>
